### ISSUE TITLE
https://firebase.google.com/codelabs/firebase-get-to-know-flutter?authuser=0#4 error on step 5

### ISSUE THREAD

**Issue by guvenacar** [flutter/flutter#142866]
I am getting provider error the end of step 5 of this tutorial (Add RSVP functionality). 
https://firebase.google.com/codelabs/firebase-get-to-know-flutter?authuser=0#4  

error code:

```
Launching lib\main.dart on Chrome in debug mode...
This app is linked to the debug service: ws://127.0.0.1:59701/r_M_M0NpAZc=/ws
Debug service listening on ws://127.0.0.1:59701/r_M_M0NpAZc=/ws
Connecting to VM Service at ws://127.0.0.1:59701/r_M_M0NpAZc=/ws
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following ProviderNotFoundException was thrown building Consumer<ApplicationState>(dirty):
Error: Could not find the correct Provider<ApplicationState> above this Consumer<ApplicationState>
Widget

This happens because you used a `BuildContext` that does not include the provider
of your choice. There are a few common scenarios:

- You added a new provider in your `main.dart` and performed a hot-reload.
  To fix, perform a hot-restart.

- The provider you are trying to read is in a different route.

  Providers are "scoped". So if you insert of provider inside a route, then
  other routes will not be able to access that provider.

- You used a `BuildContext` that is an ancestor of the provider you are trying to read.

  Make sure that Consumer<ApplicationState> is under your MultiProvider/Provider<ApplicationState>.
  This usually happens when you are creating a provider and trying to read it immediately.

  For example, instead of:

  Widget build(BuildContext context) {
    return Provider<Example>(
      create: (_) => Example(),
      // Will throw a ProviderNotFoundError, because `context` is associated
      // to the widget that is the parent of `Provider<Example>`
      child: Text(context.watch<Example>().toString()),
    );
  }

  consider using `builder` like so:

  Widget build(BuildContext context) {
    return Provider<Example>(
      create: (_) => Example(),
      // we use `builder` to obtain a new `BuildContext` that has access to the provider
      builder: (context, child) {
        // No longer throws
        return Text(context.watch<Example>().toString());
      }
    );
  }


If none of these solutions work, consider asking for help on StackOverflow:
https://stackoverflow.com/questions/tagged/flutter

The relevant error-causing widget was:
  Consumer<ApplicationState>
  Consumer:file:///D:/flutter_projeler/flutter-codelabs/firebase-get-to-know-flutter/step_02/lib/home_page.dart:26:11

When the exception was thrown, this was the stack:
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 294:3       throw_
packages/provider/src/provider.dart 343:7                                         _inheritedElementOf
packages/provider/src/provider.dart 293:30                                        of
packages/provider/src/consumer.dart 181:16                                        buildWithChild
packages/nested/nested.dart 259:41                                                build
packages/flutter/src/widgets/framework.dart 5541:22                               build
packages/nested/nested.dart 279:18                                                build
packages/flutter/src/widgets/framework.dart 5471:15                               performRebuild
packages/flutter/src/widgets/framework.dart 5187:7                                rebuild
packages/flutter/src/widgets/framework.dart 2895:18                               buildScope
packages/flutter/src/widgets/binding.dart 984:9                                   drawFrame
packages/flutter/src/rendering/binding.dart 457:5                                 [_handlePersistentFrameCallback]
packages/flutter/src/scheduler/binding.dart 1325:7                                [_invokeFrameCallback]
packages/flutter/src/scheduler/binding.dart 1255:9                                handleDrawFrame
packages/flutter/src/scheduler/binding.dart 1113:5                                [_handleDrawFrame]
lib/_engine/engine/platform_dispatcher.dart 1274:5                                invoke
lib/_engine/engine/platform_dispatcher.dart 248:5                                 invokeOnDrawFrame
lib/_engine/engine/initialization.dart 186:36                                     <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 574:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 579:39  dcall

════════════════════════════════════════════════════════════════════════════════════════════════════
Application finished.

Exited (-1).
```

**Comment by dam-ease** [comment]
Hello @guvenacar.
From the error logs, can you check this out in your existing code and make the changes.

```dart
Make sure that Consumer<ApplicationState> is under your MultiProvider/Provider<ApplicationState>.
  This usually happens when you are creating a provider and trying to read it immediately.

  For example, instead of:

  Widget build(BuildContext context) {
    return Provider<Example>(
      create: (_) => Example(),
      // Will throw a ProviderNotFoundError, because `context` is associated
      // to the widget that is the parent of `Provider<Example>`
      child: Text(context.watch<Example>().toString()),
    );
  }

  consider using `builder` like so:

  Widget build(BuildContext context) {
    return Provider<Example>(
      create: (_) => Example(),
      // we use `builder` to obtain a new `BuildContext` that has access to the provider
      builder: (context, child) {
        // No longer throws
        return Text(context.watch<Example>().toString());
      }
    );
  }
``` 

Your question is best suited for support channels like StackOverflow, so please ask the same there for solutions if any. Please also see [flutterhelp](https://www.reddit.com/r/flutterhelp/) for resources and ask questions like this.
Closing from here as this is more of a how-to question rather than a bug or feature request.
If you are proposing a change, please use the appropriate new issue template and fill out the requested details so that we can address this issue properly.

**Comment by guvenacar** [comment]
Actually, I know I shouldn't ask this question here. What I really want to point out is that the flutter official tutorial has a bug in step 5. Many people like me who want to learn flutter will try this but they will get errors. It is unacceptable for this error to be in the tutorial.

**Comment by dam-ease** [comment]
Hello @guvenacar
Please can you confirm you cloned the repo as stated here firstly https://firebase.google.com/codelabs/firebase-get-to-know-flutter?authuser=0#1
And can you help with information on if this occurs only after step 5 and not on initial run of the cloned repo?

**Comment by guvenacar** [comment]
Hello [dam-ease](https://github.com/dam-ease)
Yes this time worked well, thank you. 
Actulally i was copied manually "flutter-codelabs" directory from github and i had do same things but it was didn't work. Perhaps i need use github clone code?!
But even this way is difficult for new users.
From what I understand there have been some fundamental changes to flutter. Therefore, none of the solutions I found on the internet worked. All are old examples. I think you should make a clear and simple tutorial for new users with the new flutter's fundamental , without involving github. 
Sorry for wasting your time.

**Comment by guvenacar** [comment]
That's incredible. I'm going crazy. After saying that it worked after step 5, it gave an error again in step 8. I tried copying from github and running the project again several times, but it gave error again at step 5.
Again I can't get past step 5. When the following code is added to the hope_page.dart file, it gives an error.
          // Add from here
           Consumer<ApplicationState>(
             builder: (context, appState, _) => AuthFunc(
                 loggedIn: appState.loggedIn,
                 signOut: () {
                   FirebaseAuth.instance.signOut();
                 }),
           ),
           // to here

**Comment by dam-ease** [comment]
Hello @guvenacar.
From the logs, this might be an issue with your implementation. 
You can find solutions to these and ask questions in support channels like StackOverflow or  [flutterhelp](https://www.reddit.com/r/flutterhelp/). If the code sample gives errors as a result of codes in the codelabs, please provide links to the code

**Comment by guvenacar** [comment]
Thank you for your interest. I found the solution based on the answer to the question I asked on Stackoverflow.

https://stackoverflow.com/questions/77969502/i-cant-pass-step-5-of-get-to-know-firebase-for-flutter-tutorial/77969997#77969997

**Comment by guvenacar** [comment]
Hello [dam-ease](https://github.com/dam-ease) 
in tutorial step 9, in app_state.dart file there is no declare  _emailVerified then getting error. 
I added these lines
 bool _emailVerified = false; 
 bool get emailVerified => _emailVerified; 
app_state.dart page in tutorial should be like this:

`
import 'dart:async';                                     // new

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart'
    hide EmailAuthProvider, PhoneAuthProvider;
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_ui_auth/firebase_ui_auth.dart';
import 'package:flutter/material.dart';

import 'firebase_options.dart';
import 'guest_book_message.dart';                        // new

enum Attending { yes, no, unknown }

class ApplicationState extends ChangeNotifier {
  ApplicationState() {
    init();
  }

int _attendees = 0;
int get attendees => _attendees;

Attending _attending = Attending.unknown;
StreamSubscription<DocumentSnapshot>? _attendingSubscription;
Attending get attending => _attending;
set attending(Attending attending) {
  final userDoc = FirebaseFirestore.instance
      .collection('attendees')
      .doc(FirebaseAuth.instance.currentUser!.uid);
  if (attending == Attending.yes) {
    userDoc.set(<String, dynamic>{'attending': true});
  } else {
    userDoc.set(<String, dynamic>{'attending': false});
  }
}

  bool _loggedIn = false;
  bool get loggedIn => _loggedIn; 

  //Declare emailVerified here
  bool _emailVerified = false; 
  bool get emailVerified => _emailVerified; 
  
  // Add from here...
  StreamSubscription<QuerySnapshot>? _guestBookSubscription;
  List<GuestBookMessage> _guestBookMessages = [];
  List<GuestBookMessage> get guestBookMessages => _guestBookMessages;
  // ...to here.

  Future<void> init() async {
    await Firebase.initializeApp(
        options: DefaultFirebaseOptions.currentPlatform);

    FirebaseUIAuth.configureProviders([
      EmailAuthProvider(),
    ]);

    // Add from here...
    FirebaseFirestore.instance
        .collection('attendees')
        .where('attending', isEqualTo: true)
        .snapshots()
        .listen((snapshot) {
      _attendees = snapshot.docs.length;
      notifyListeners();
    });
    // ...to here.

    FirebaseAuth.instance.userChanges().listen((user) {
      if (user != null) {
        _loggedIn = true;
        _guestBookSubscription = FirebaseFirestore.instance
            .collection('guestbook')
            .orderBy('timestamp', descending: true)
            .snapshots()
            .listen((snapshot) {
          _guestBookMessages = [];
          for (final document in snapshot.docs) {
            _guestBookMessages.add(
              GuestBookMessage(
                name: document.data()['name'] as String,
                message: document.data()['text'] as String,
              ),
            );
          }
          notifyListeners();
        });
        // Add from here...
        _attendingSubscription = FirebaseFirestore.instance
            .collection('attendees')
            .doc(user.uid)
            .snapshots()
            .listen((snapshot) {
          if (snapshot.data() != null) {
            if (snapshot.data()!['attending'] as bool) {
              _attending = Attending.yes;
            } else {
              _attending = Attending.no;
            }
          } else {
            _attending = Attending.unknown;
          }
          notifyListeners();
        });
        // ...to here.

      } else {
        _loggedIn = false;
        _emailVerified = false;
        _guestBookMessages = [];
        _guestBookSubscription?.cancel();
        _attendingSubscription?.cancel(); // new
      }
      notifyListeners();
    });


    
  }

   // Add from here...
  Future<DocumentReference> addMessageToGuestBook(String message) {
    if (!_loggedIn) {
      throw Exception('Must be logged in');
    }

    return FirebaseFirestore.instance
        .collection('guestbook')
        .add(<String, dynamic>{
      'text': message,
      'timestamp': DateTime.now().millisecondsSinceEpoch,
      'name': FirebaseAuth.instance.currentUser!.displayName,
      'userId': FirebaseAuth.instance.currentUser!.uid,
    });
  }
  // ...to here.
} 
`
